WalletWasabi/WabiSabi/Backend/Rounds/Arena.cs

```diff
@@ -509,34 +511,39 @@ public async Task<ReissueCredentialResponse> ReissuanceAsync(ReissueCredentialRe
 				}
+			}
 
-				if (request.RealAmountCredentialRequests.Delta != 0)
-				{
-					throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.DeltaNotZero, $"Round ({round.Id}): Amount credentials delta must be zero.");
-				}
-
-				if (request.RealAmountCredentialRequests.Requested.Count() != ProtocolConstants.CredentialNumber)
-				{
-					throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.WrongNumberOfCreds, $"Round ({round.Id}): Incorrect requested number of amount credentials.");
-				}
+			if (request.RealAmountCredentialRequests.Delta != 0)
+			{
+				throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.DeltaNotZero, $"Round ({round.Id}): Amount credentials delta must be zero.");
+			}
 
-				if (request.RealVsizeCredentialRequests.Requested.Count() != ProtocolConstants.CredentialNumber)
-				{
-					throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.WrongNumberOfCreds, $"Round ({round.Id}): Incorrect requested number of weight credentials.");
-				}
+			if (request.RealAmountCredentialRequests.Requested.Count() != ProtocolConstants.CredentialNumber)
+			{
+				throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.WrongNumberOfCreds, $"Round ({round.Id}): Incorrect requested number of amount credentials.");
+			}
 
-				var commitRealAmountCredentialResponse = round.AmountCredentialIssuer.PrepareResponse(request.RealAmountCredentialRequests);
-				var commitRealVsizeCredentialResponse = round.VsizeCredentialIssuer.PrepareResponse(request.RealVsizeCredentialRequests);
-				var commitZeroAmountCredentialResponse = round.AmountCredentialIssuer.PrepareResponse(request.ZeroAmountCredentialRequests);
-				var commitZeroVsizeCredentialResponse = round.VsizeCredentialIssuer.PrepareResponse(request.ZeroVsizeCredentialsRequests);
-
-				return new(
-					commitRealAmountCredentialResponse.Commit(),
-					commitRealVsizeCredentialResponse.Commit(),
-					commitZeroAmountCredentialResponse.Commit(),
-					commitZeroVsizeCredentialResponse.Commit()
-					);
+			if (request.RealVsizeCredentialRequests.Requested.Count() != ProtocolConstants.CredentialNumber)
+			{
+				throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.WrongNumberOfCreds, $"Round ({round.Id}): Incorrect requested number of weight credentials.");
 			}
+
+			var realAmountTask = Task.Run(() => round.AmountCredentialIssuer.PrepareResponse(request.RealAmountCredentialRequests), cancellationToken);
+			var realVsizeTask = Task.Run(() => round.VsizeCredentialIssuer.PrepareResponse(request.RealVsizeCredentialRequests), cancellationToken);
+			var zeroAmountTask = Task.Run(() => round.AmountCredentialIssuer.PrepareResponse(request.ZeroAmountCredentialRequests), cancellationToken);
+			var zeroVsizeTask = Task.Run(() => round.VsizeCredentialIssuer.PrepareResponse(request.ZeroVsizeCredentialsRequests), cancellationToken);
+
+			var commitRealAmountCredentialResponse = await realAmountTask.ConfigureAwait(false);
```

Comment:
At this point you don't know whether you are in `output registration` phase or not.

Reply:
What is the consequence?

Reply:
This is fine, because output registration is the last phase during which credential presentations can change the transaction.

If clients are honest, this can only happen during a timeout, because an honest client would not signal `ReadyToSign` while it still has reissuance requests in flight.

Reply:
I was just pointing out the change in behavior. Previously to this PR the validation made sure the phase is `ConnectionConfirmation` or `OutputRegistration` while now, that validation is still in place but it could happen that the reissuance happens anyway even if not in any of those phases.

Reply:
Yep, changing this behavior was a conscious decision.

That check still before here same as before:

https://github.com/zkSNACKs/WalletWasabi/blob/c9adc201390869a8b95485b782a927f063814bbf/WalletWasabi/WabiSabi/Backend/Rounds/Arena.cs#L510-L513

so the phase cannot be input registration, it can only be one of these phases, or if made by a malicious client, it can also be the signing phase by the end. However this gives malicious clients no additional advantage.

We could recheck the phase, but if it's no longer valid there's nothing the coordinator can do with this information, there's there's no associated input that can be noted at this point, and the returned credentials can't be used to do anything else, so handling this edge case would make the code more complex without any clear advantage.